File: src/backend/studio-reservation/src/main/java/org/taonaw/studio_reservation/domain/model/memberAccount/MemberAccount.java
package org.taonaw.studio_reservation.domain.model.memberAccount;

import lombok.NonNull;
import org.taonaw.studio_reservation.domain.model.shared.*;

import java.util.Objects;

public class MemberAccount {
    private MemberAccountId id;
    private PersonName name;
    private EmailAddress emailAddress;
    private HashedPassword password;
    private DateOfBirth dateOfBirth;
    private PhoneNumber phoneNumber;

    public boolean authenticate(
            @NonNull EmailAddress emailAddress,
            @NonNull PlainTextPassword plainTextPassword,
            @NonNull PasswordEncoder passwordEncoder) {

        return this.emailAddress.equals(emailAddress) && plainTextPassword.matches(this.password, passwordEncoder);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        MemberAccount that = (MemberAccount) o;
        return id.equals(that.id);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id);
    }
}

File: src/backend/identityaccess/src/main/kotlin/org/taonaw/identityaccess/domain/model/memberAccount/MemberAccount.kt
package org.taonaw.identityaccess.domain.model.memberAccount

import org.taonaw.identityaccess.domain.model.shared.HashedPassword
import org.taonaw.identityaccess.domain.model.shared.PasswordHashingService
import org.taonaw.identityaccess.domain.model.shared.PlainTextPassword

class MemberAccount private constructor(
        val memberAccountId: MemberAccountId,
        val memberName: MemberName,
        val contractInformation: ContractInformation,
        val password: HashedPassword) {

    companion object {
        fun create(
                memberName: MemberName,
                contractInformation: ContractInformation,
                plainTextPassword: PlainTextPassword,
                passwordHashingService: PasswordHashingService): MemberAccount {

            return MemberAccount(MemberAccountId.newId(), memberName, contractInformation, plainTextPassword.hash(passwordHashingService))
        }
    }

    fun change(memberName: MemberName, contractInformation: ContractInformation): MemberAccount {
        return MemberAccount(memberAccountId, memberName, contractInformation, password)
    }

    fun changePassword(
            oldPlainTextPassword: PlainTextPassword,
            newPlainTextPassword: PlainTextPassword,
            passwordHashingService: PasswordHashingService): MemberAccount {

        if (!oldPlainTextPassword.matches(password, passwordHashingService))
            MemberAccountOldPasswordDifferentErr().throwErr()

        return MemberAccount(memberAccountId, memberName, contractInformation, newPlainTextPassword.hash(passwordHashingService))
    }

    fun authenticate(plainTextPassword: PlainTextPassword, passwordHashingService: PasswordHashingService): Boolean {
        return plainTextPassword.matches(password, passwordHashingService)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as MemberAccount

        if (memberAccountId != other.memberAccountId) return false

        return true
    }

    override fun hashCode(): Int {
        return memberAccountId.hashCode()
    }
}